lib/hawkbit/hawkbit.c

```diff
@@ -0,0 +1,1273 @@
+		close.status.result.finished = fini;
+        	ret = json_obj_encode_buf(json_close_descr,
+				ARRAY_SIZE(json_close_descr),
+                                &close, hb_context.status_buffer,
+                		hb_context.status_buffer_size -1);
+		if (ret)  {
+			LOG_ERR("Can't encode the json script");
+			return false;
+		}
+
+		hb_context.http_req.content_type_value =
+			HTTP_HEADER_CONTENT_TYPE_JSON;
+		hb_context.http_req.payload =
+			hb_context.status_buffer;
+		hb_context.http_req.payload_len =
+			strlen(hb_context.status_buffer);
+
+	        ret = http_client_req(hb_context.sock, &hb_context.http_req,
+                             HAWKBIT_RECV_TIMEOUT, HAWKBIT_CLOSE);
+		if (ret < 0) {
+			LOG_ERR("Unable to send http request");
+			return false;
+        	}
+		break;
+
+	case HAWKBIT_PROBE_DEPLOYMENT_BASE:
+		hb_context.http_req.content_type_value = NULL;
+		ret = http_client_req(hb_context.sock, &hb_context.http_req,
+				      HAWKBIT_RECV_TIMEOUT, HAWKBIT_PROBE_DEPLOYMENT_BASE);
+		if (ret < 0) {
+			LOG_ERR("Unable to send http request");
+			return false;
+		}
+		break;
+
+	case HAWKBIT_REPORT:
+		if (!fini || !exec) {
+			return -EINVAL;
+		}
+
+		LOG_INF("Reporting deployment feedback %s (%s) for action %d",
+				fini, exec, json_action_id);
+		/* Build JSON */
+		memset(&feedback, 0, sizeof(feedback));
+		snprintk(acid, sizeof(acid), "%d", json_action_id);
+		feedback.id = acid;
+		feedback.status.result.finished = fini;
+		feedback.status.execution = exec;
+
+		ret = json_obj_encode_buf(json_dep_fbk_descr,
+				ARRAY_SIZE(json_dep_fbk_descr),
+				&feedback,  hb_context.status_buffer,
+                                hb_context.status_buffer_size -1);
+		if (ret) {
+			LOG_ERR("Can't encode response: %d", ret);
+			return ret;
+		}
+
+		hb_context.http_req.content_type_value =
+                        HTTP_HEADER_CONTENT_TYPE_JSON;
+                hb_context.http_req.payload =
+                        hb_context.status_buffer;
+                hb_context.http_req.payload_len =
+                        strlen(hb_context.status_buffer);
+
+                ret = http_client_req(hb_context.sock,
+				      &hb_context.http_req,
+				      HAWKBIT_RECV_TIMEOUT, HAWKBIT_REPORT);
+
+                if (ret < 0) {
+                        LOG_ERR("Unable to send http request");
+                        return false;
+                }
+		break;
+
+	case HAWKBIT_DOWNLOAD:
+		ret = http_client_req(hb_context.sock,
+				      &hb_context.http_req,
+                                      HAWKBIT_RECV_TIMEOUT, HAWKBIT_DOWNLOAD);
+                if (ret < 0) {
+                        LOG_ERR("Unable to send image download request");
+                        return false;
+                }
+		break;
+	}
+
+	k_free(device_id);
+	return true;
+}
+
+
+enum hawkbit_response hawkbit_probe(void)
+{
+	int ret;
+	s32_t file_size = 0;
+	struct hawkbit_device_acid device_action_id;
+
+	char *device_id = k_malloc(DEVICE_ID_MAX_SIZE);
```

Comment:
Using k_malloc() should be avoided (in a code inside zephyr) unless there is no other alternative. Is it possible to avoid using k_malloc here, could you let the caller to allocate space for these?